spatialos-sdk/src/worker/component.rs

```diff
@@ -1,35 +1,463 @@
+        pub user_handle: *mut Worker_CommandResponseHandle,
+    }
+
+    impl From<&Worker_CommandResponse> for CommandResponse {
+        fn from(response: &Worker_CommandResponse) -> Self {
+            CommandResponse {
+                component_id: response.component_id,
+                schema_type: SchemaCommandResponse {
+                    component_id: response.component_id,
+                    internal: response.schema_type,
+                },
+                user_handle: response.user_handle,
+            }
         }
     }
 }
 
-// TODO: Wrap Schema_ComponentUpdate
-pub struct ComponentUpdate {
-    pub component_id: ComponentId,
-    pub schema_type: *mut Schema_ComponentUpdate,
+// A data structure which represents all known component types. Used to generate an array of vtables to pass
+// to the connection object.
+pub struct ComponentDatabase {
+    component_vtables: Vec<worker::Worker_ComponentVtable>,
+}
+
+impl ComponentDatabase {
+    pub fn new() -> Self {
+        ComponentDatabase {
+            component_vtables: Vec::new(),
+        }
+    }
+
+    pub fn add_component<C: Component>(mut self) -> Self {
+        self.component_vtables.push(create_component_vtable::<C>());
+        self
+    }
+
+    pub(crate) fn to_worker_sdk(&self) -> *const worker::Worker_ComponentVtable {
+        self.component_vtables.as_ptr()
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.component_vtables.len()
+    }
+}
+
+// An object which contains a reference counted instance to T.
+struct ClientHandle<T> {
+    data: Rc<T>,
+    // When Drop() is called, data should reduce its reference count.
 }
 
-impl From<&Worker_ComponentUpdate> for ComponentUpdate {
-    fn from(update: &Worker_ComponentUpdate) -> Self {
-        ComponentUpdate {
-            component_id: update.component_id,
-            schema_type: update.schema_type,
+impl<T> ClientHandle<T> {
+    fn new(data: T) -> ClientHandle<T> {
+        ClientHandle {
+            data: Rc::new(data),
+        }
+    }
+
+    fn copy(&self) -> ClientHandle<T> {
+        ClientHandle {
+            data: Rc::clone(&self.data),
         }
     }
+
+    // Vtable helper functions.
+    unsafe fn handle_allocate(data: T) -> *mut raw::c_void {
+        // Allocate client handle and initialize with data.
+        let new_client_handle: *mut ClientHandle<T> =
+            mem::transmute(alloc::alloc(Layout::new::<ClientHandle<T>>()));
+        ptr::write(new_client_handle, ClientHandle::<T>::new(data));
+
+        // Return new pointer.
+        mem::transmute(new_client_handle)
+    }
+
+    unsafe fn handle_free(handle: *mut raw::c_void) {
+        let client_handle_ptr: *mut ClientHandle<T> = mem::transmute(handle);
+
+        // Call drop() on pointer value.
+        client_handle_ptr.drop_in_place();
+
+        // Deallocate memory.
+        alloc::dealloc(
+            mem::transmute(client_handle_ptr),
+            Layout::new::<ClientHandle<T>>(),
+        );
```

Comment:
I think this function could be simplified down to

```
let _ = Arc::<T>::from_raw(handle);
```

That will reconstruct the `Arc` from the raw pointer and then drop it, decrementing the counter (and thereby freeing the data if it was the last handle).

Reply:
Neat. Agreed.